src/v/cloud_storage/tx_range_manifest.cc

```diff
@@ -12,20 +12,239 @@
+            return true;
+        case state::expect_range_first_value:
+            ranges.back().first = model::offset(i64);
+            _range_keys |= static_cast<uint8_t>(range_keys::first);
+            _state = state::expect_range_key;
+            return true;
+        case state::expect_range_last_value:
+            ranges.back().last = model::offset(i64);
+            _range_keys |= static_cast<uint8_t>(range_keys::last);
+            _state = state::expect_range_end_object;
+            return true;
+        default:
+            return false;
+        }
+    }
+    bool Uint64(uint64_t u64) {
+        if (u64 > std::numeric_limits<int64_t>::max()) {
+            return false;
+        } else {
+            return Int64(static_cast<int64_t>(u64));
+        }
+    }
+    bool Double(double) { return false; }
+    /// enabled via kParseNumbersAsStringsFlag, string is not null-terminated
+    /// (use length)
+    bool
+    RawNumber(const Ch* /*str*/, rapidjson::SizeType /*len*/, bool /*copy*/) {
+        return false;
+    }
+    bool String(const Ch*, rapidjson::SizeType, bool) { return false; }
+    bool StartObject() {
+        switch (_state) {
+        case state::expect_manifest_object:
+            _state = state::expect_manifest_key;
+            return true;
+        case state::expect_ranges_entry_or_end_array:
+            ranges.push_back(model::tx_range());
+            _range_keys = 0;
+            _state = state::expect_range_key;
+            return true;
+        default:
+            return false;
+        }
+    }
+    bool Key(const Ch* str, rapidjson::SizeType len, bool /*copy*/) {
+        switch (_state) {
+        case state::expect_manifest_key:
+            if (std::string_view(str, len) == "version") {
+                _state = state::expect_version_value;
+                return true;
+            } else if (std::string_view(str, len) == "compat_version") {
+                _state = state::expect_compat_version_value;
+                return true;
+            } else if (std::string_view(str, len) == "ranges") {
+                _state = state::expect_ranges_value;
+                return true;
+            } else {
+                return false;
+            }
+        case state::expect_range_key:
+            if (std::string_view(str, len) == "pid.id") {
+                _state = state::expect_range_pid_id_value;
+                return true;
+            } else if (std::string_view(str, len) == "pid.epoch") {
+                _state = state::expect_range_pid_epoch_value;
+                return true;
+            } else if (std::string_view(str, len) == "first") {
+                _state = state::expect_range_first_value;
+                return true;
+            } else if (std::string_view(str, len) == "last") {
+                _state = state::expect_range_last_value;
+                return true;
+            } else {
+                return false;
+            }
+        default:
```

Comment:
nit: I prefer to list out all the other cases here; the compiler then tells me when I added a state I forgot to think about.

Reply:
yeh, i missed this. good catch

Reply:
I usually agree with this but here I prefer to make an exception. The rule is good when you are supposed to handle all/most keys everywhere. Here, keys are used only once. Hard to forget to use it. Also it is all private.

tl;dr; too much unnecessary code for little to no benefit